lib/util/github/graphql/cache-adapters/abstract-cache-adapter.ts

```diff
@@ -0,0 +1,156 @@
+    if (storedData) {
+      const { cacheTTLDays } = AbstractGithubGraphqlCacheAdapter;
+      const cacheTTLDuration = { days: cacheTTLDays };
+      if (!isDateExpired(this.now, storedData.createdAt, cacheTTLDuration)) {
+        result = storedData;
+      }
+    }
+
+    this.createdAt = DateTime.fromISO(result.createdAt);
+    this.updatedAt = DateTime.fromISO(result.updatedAt);
+    this.items = result.items;
+    return this.items;
+  }
+
+  /**
+   * If a package version was published less than `packageUnstableDays` ago,
+   * then it is considered unstable and can be evicted from cache.
+   *
+   * Otherwise, it is considered stabilized and will be persisted until
+   * cache record expires.
+   */
+  private isStabilized(item: GithubItem): boolean {
+    const { packageUnstableDays } = AbstractGithubGraphqlCacheAdapter;
+    const unstableDuration = { days: packageUnstableDays };
+    return isDateExpired(this.now, item.releaseTimestamp, unstableDuration);
+  }
+
+  /**
+   * Process items received from GraphQL page
+   * ordered by `releaseTimestamp` in descending order
+   * (fresh versions go first).
+   */
+  async reconcile(items: GithubItem[]): Promise<boolean> {
+    const cachedItems = await this.getItems();
+
+    let isPaginationDone = false;
+    for (const item of items) {
+      const { version } = item;
+
+      // If we reached previously stored item that is stabilized,
+      // we assume the further pagination will not yield any new items.
+      const oldItem = cachedItems[version];
+      if (oldItem && this.isStabilized(oldItem)) {
+        isPaginationDone = true;
+        break;
+      }
+
+      cachedItems[version] = item;
+      this.reconciledVersions.add(version);
+    }
+
+    this.items = cachedItems;
+    return isPaginationDone;
+  }
+
+  /**
+   * Handle removed items for packages that are not stabilized
+   * and return the list of all items.
+   */
+  async finalize(): Promise<GithubItem[]> {
+    const cachedItems = await this.getItems();
+    const resultItems: Record<string, GithubItem> = {};
+
+    for (const [version, item] of Object.entries(cachedItems)) {
+      if (this.isStabilized(item) || this.reconciledVersions.has(version)) {
+        resultItems[version] = item;
+      }
+    }
+
+    await this.store(resultItems);
+    return Object.values(resultItems);
+  }
+
+  /**
+   * Update `updatedAt` field and persist the data.
+   */
+  private async store(cachedItems: Record<string, GithubItem>): Promise<void> {
+    const cacheRecord: GithubGraphqlCacheRecord<GithubItem> = {
+      items: cachedItems,
+      createdAt: this.createdAt.toISO(),
+      updatedAt: this.now.toISO(),
+    };
+    await this.persist(cacheRecord);
+  }
+
+  /**
+   * Loading and persisting data is delegated to the concrete adapter.
+   */
+  abstract load(): Promise<GithubGraphqlCacheRecord<GithubItem> | undefined>;
+  abstract persist(_: GithubGraphqlCacheRecord<GithubItem>): Promise<void>;
```

Comment:
```suggestion
  abstract persist(cacheRecord: GithubGraphqlCacheRecord<GithubItem>): Promise<void>;
```
we should use full name and reuse in overrides to be consistent